Add config system to Quilt Loader

Haven-King opened the pull request:
This PR adds several features to Quilt Loader:

- A low-level system for creating, grouping, tracking, and serializing values that can be used to create higher-level config systems
- Config values, sections, and files can each have metadata associated with them
- A system to register serializers for various file types
- A feature to set a global serializer to be used by all config files
- A feature to set the default serializer to be used by each config file that does not explicitly declare a specific file format
- Built-in json5 and TOML serializers, with config files being TOML by default
- A high-level system for registering and accessing config values (see [ConfigTester#testReflectiveConfigs](https://github.com/Haven-King/quilt-loader/blob/41238871c5a73fa77e3b7920dae5a4fb14a27f01/src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java#L153) and [TestReflectiveConfig](https://github.com/Haven-King/quilt-loader/blob/master/src/test/java/org/quiltmc/loader/impl/config/TestReflectiveConfig.java))

A majority of the contents of this PR were moved to a separate library, [QuiltConfig](https://github.com/QuiltMC/quilt-config).

### Why put this in Loader and not QSL?
So that it could be used in other parts of Loader and so that mods can have configs independent of whether or not QSL is present.

### What about file format XYZ?
TOML was chosen as the default because it is the best-looking format for configs with a low level of complexity, which covers most configs. It also has reasonably good library support, which made it a much more appealing choice than YAML. For more complex configs, a json5 serializer has been included that mod authors can opt to use instead.

**Please keep all discussion related to config formats out of this thread.** There has been ample discussion on the matter [here](https://github.com/QuiltMC/rfcs/issues/36). You are welcome to comment there if you have any really compelling arguments for another file format.

Haven-King committed:
Initial commit of config system

Haven-King committed:
Migrate existing configs from their default types to the global type if a global config file type exists

Haven-King committed:
Abstract config save location somewhat from serialization logic

TheGlitch76 commented:
Well, that came out of left field.

OroArmor reviewed (commented):
Looks good from an initial review. I would add some more docs and make some deeper impl packages for clarity

Haven-King committed:
- Reorganize packages
- Add regex
- Actually store created configs

Leo40Git commented on src/main/java/org/quiltmc/loader/impl/config/Json5Serializer.java:
```diff
@@ -0,0 +1,260 @@
+					if (m.containsKey(k) && i != value.getKey().length() - 1) {
+						m = (Map<String, Object>) m.get(k);
+					} else if (m.containsKey(k)) {
+						//noinspection rawtypes
```
`noinspection` comments are exclusive to IDEA. This should be moved to a `@SuppressWarnings` annotation.

Leo40Git commented on src/main/java/org/quiltmc/loader/impl/config/builders/ReflectiveConfigCreatorImpl.java:
```diff
@@ -0,0 +1,105 @@
+						try {
+							field.setAccessible(true);
+							field.set(object, newValue);
+							field.setAccessible(false);
```
Why reset the accessible flag?

Leo40Git commented on src/main/java/org/quiltmc/loader/impl/config/tree/Trie.java:
```diff
@@ -0,0 +1,263 @@
+import org.quiltmc.loader.impl.config.builders.SectionBuilderImpl;
+import org.quiltmc.loader.impl.config.values.ValueKeyImpl;
+
+public final class Trie {
```
Is this intentionally misspelled?

Haven-King replied:
It is not misspelled. A [Trie](https://en.wikipedia.org/wiki/Trie) or prefix tree is its own data structure.

Leo40Git replied:
Oh, oops. My mistake, I have never heard of tries.

Leo40Git reviewed (approved):
Seems good, beyond the lack of javadocs. Just a few nitpicks...

Haven-King committed:
Replace noinspection comments with @SuppressWarnings

Haven-King committed:
Call update listeners whenever an override is added or removed

Haven-King committed:
Move field accessibility modifier outside of lambda and don't revert it

Haven-King committed:
- Add lots of javadoc
- Add some safety mechanisms

Haven-King commented:
> Looks good from an initial review. I would add some more docs and make some deeper impl packages for clarity

Packages have been reorganized and a lot more JavaDocs have been added. Since as an API developer it can sometimes be hard to look at it from an objective view. As such, please recommend any other docs that may be needed.

osfanbuff63 commented:
License violations? o_o

Haven-King committed:
Allow configuration of the default config file type

Haven-King committed:
Add enum support

Haven-King committed:
Specify comment target

eristhea reviewed (approved):
Really nitpicky review for javadoc grammar!
Overall this looks really nice, thank you for working on it

Haven-King committed:
Apply grammar suggestions from code review

Co-authored-by: ix0rai <ix0rai64@gmail.com>

Haven-King committed:
Add javadoc to Config.Builder#fileType

Haven-King committed:
Merge remote-tracking branch 'origin/master'

Haven-King committed:
Add TOML support and make TOML the default config format

osfanbuff63 commented:
Is there a reason build is failing with `org.cadixdev.gradle.licenser.LicenseViolationException`?

wafflecoffee commented:
@osfanbuff63 Haven forgot to apply license headers. It's not something to be worried about.

Haven-King commented:
> @osfanbuff63 Haven forgot to apply license headers. It's not something to be worried about.

The license checker only accepts FabricMC licenses right now. As this is obviously original Quilt code, I used QuiltMC license headers, which do not pass the current check. That seems like something that should be fixed upstream.

i509VCB commented on build.gradle:
```diff
@@ -80,6 +80,8 @@ dependencies {
 	include 'org.quiltmc:quilt-loader-sat4j:2.3.5.1'
 
+	api 'com.electronwill.night-config:core:3.6.5'
+	api 'com.electronwill.night-config:toml:3.6.5'
```
New dependencies need to be added the json files for launcher dependencies.

Haven-King committed:
Remove redundant initialization call

osfanbuff63 commented:
> The license checker only accepts FabricMC licenses right now. As this is obviously original Quilt code, I used QuiltMC license headers, which do not pass the current check. That seems like something that should be fixed upstream.

Upstream = fabric? Or what else? (Just confused :P)

Haven-King commented:
Upstream from my fork, so Quilt Loader.

wafflecoffee commented:
> The license checker only accepts FabricMC licenses right now. As this is obviously original Quilt code, I used QuiltMC license headers, which do not pass the current check. That seems like something that should be fixed upstream.

Cunningham's Law strikes again!

CheaterCodes reviewed (commented):
Well, this took me a while to get done...
I looked very closed and picked lots of nits.

Looks great in general though!

Kroppeb commented on src/main/java/org/quiltmc/loader/api/config/Constraint.java:
```diff
@@ -0,0 +1,98 @@
+		};
+	}
+
+	class Range<T extends Number> implements Constraint<T> {
```
I guess `Comparable<T>` could work, but I think having string ranges is a bit weird?

Kroppeb replied:
ugh, github only shows the code above the comment, given that there is already a comparator being passed, you wouldn't need to have any bound at all

CheaterCodes replied:
I agree that it's unnecessary to abstract this, but the Number constraint here is completely arbitrary and unnecessary.
If it was any effort to support other comparables, I wouldn't bother, but I don't see why you spend extra effort restricting this.

Kroppeb commented on src/main/java/org/quiltmc/loader/impl/config/ConfigImpl.java:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	public Iterable<TrackedValue<?>> values() {
+		return new Iterable<TrackedValue<?>>() {
```
Yeah, next should call `hasNext()` and throw a `NoSuchEllementException` if it returns false

CheaterCodes commented on src/main/java/org/quiltmc/loader/api/config/MetadataType.java:
```diff
@@ -0,0 +1,31 @@
+ */
+@ApiStatus.NonExtendable
+public interface MetadataType<T> {
+	String getId();
```
I don't think that would happen often. I think raw string metadata should have a lightweight wrapper class anyway (for constants or helper methods).

Haven-King replied:
It would make sense to me to have a class passed *in addition to* the string ID, but it doesn't make sense to me to use that as identifying information. Having wrapper classes over String metadata also doesn't seem very useful to me.

CheaterCodes replied:
I think I've had the exact same argument with the attribute lookup api. I'm just a fan of using strings and magic constants when they're not needed.
I guess it mostly comes down to what metadata is used for. Other than comments.

Haven-King committed:
Fix potential NPE

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,420 @@
+	}
+
+	@ApiStatus.NonExtendable
+	interface Builder {
```
It could extend `SectionBuilder` to adhere to OOP best practices, but I'd probably leave the methods declared explicitly as well for visibility purposes, at which point it wouldn't make much of a difference.

CheaterCodes replied:
The difference is that someone could write generic code over it.

Haven-King replied:
There's also no self type in Java, so the return types of the methods would be wrong unless I did some very ugly generics or widening of the method return types.

CheaterCodes replied:
Do interfaces not support bridge methods?

Haven-King replied:
I had a type mismatch, I can totally make this happen.

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,420 @@
+	 */
+	void register(UpdateCallback callback);
+
+	Iterable<String> flags();
```
Flags are another kind of metadata. Essentially, flags are unique string values that are stored in a set, while metadata has a list of metadata values for each type. I did consider making flags just a metadata type instead of its own construct, but I thought that the uniqueness might warrant keeping them around.

CheaterCodes replied:
I think for simplicity, Flags should be metadata. However, having these existing functions as helpers could still be nice.

Haven-King replied:
I don't think the added simplicity of having one metadata type outweighs the added complexity of consumers having to check for potential duplicates.

CheaterCodes replied:
Why would consumers have to check for duplicates?
Just make a Metadata for a Set of Strings for flags. Pretty sure that's the exact same functionality.

Haven-King committed:
Have shorthand methods call the base API method instead of the base impl method

Haven-King committed:
Use default parameters for System#getProperty calls

Haven-King committed:
Small iterator changes

Haven-King committed:
Small iterator fix

Haven-King committed:
Another small iterator change

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/Serializer.java:
```diff
@@ -0,0 +1,35 @@
+
+	void serialize(Config config, OutputStream to) throws IOException;
+
+	void deserialize(Config config, InputStream from);
```
Because my default deserializers didn't, so I forgot to add it 😅

Haven-King committed:
- Allow Serializer#deserialize to throw IOExceptions
- Small fix in NightConfigSerializer

Haven-King commented on src/main/java/org/quiltmc/loader/impl/config/tree/Trie.java:
```diff
@@ -0,0 +1,263 @@
+		return oldValue;
+	}
+
+	public ValueTreeNode put(ValueKey key, ValueTreeNode value) {
```
Fair enough, this method might as well take an `Iterable` too. The argument order is reversed on the other method so that it can use varargs for the key. That method isn't even used though, so I guess I might as well remove it.

Haven-King committed:
- Remove unused methods in Trie
- Widen usages of `ValueKey` in parameter types to `Iterable<String>`

Haven-King committed:
Flatten modifications made by creating a section

Haven-King committed:
Actually trigger config update callbacks

Haven-King committed:
Make creating a ReflectiveConfigCreator use existing methods in ConfigImpl

Haven-King committed:
Move some build logic out of ConfigImpl and into ConfigBuilderImpl

Haven-King committed:
Use UncheckedIOException instead of raw RuntimeException

Haven-King committed:
Rename register to registerCallback

Haven-King committed:
Fix javadocs and add @NotNull annotation to TrackedValue params

Haven-King committed:
Minor javadoc changes

Haven-King commented on src/main/java/org/quiltmc/loader/impl/config/tree/TrackedValueImpl.java:
```diff
@@ -0,0 +1,188 @@
+		this.valueOverride = newValue;
+
+		for (UpdateCallback<T> callback : this.callbacks) {
+			callback.onUpdate(this.key, this.value, newValue);
```
I don't see why not.

CheaterCodes replied:
Well, since you pass an oldValue and a newValue, I would assume the callback is invoked whenever `oldValue != newValue`.
Furthermore, I'd assume that oldValue was the previous value, which is also not quite true here.
I presonally think this should only trigger when the *visible* value (as returned by getValue) actually changed, but if you decide to *always* call it, I would probably just omit the `oldValue` and `newValue` arguments to avoid confusion.

CheaterCodes replied:
Just make sure to have a contract that you stick with and is consistent. Otherwise the API becomes hard to understand.
I'd rather have you *always* trigger the callbacks if any of these methods are called than sometimes, according to a rule that's not clearly visible.

Haven-King committed:
Expose ConfigImpl#serialize in Config

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/TrackedValue.java:
```diff
@@ -0,0 +1,215 @@
+	 * @param serialize whether or not to serialize this values config file. Should be false only when deserializing
+	 * @return the old value that's been replaced
+	 */
+	T setValue(T newValue, boolean serialize);
```
It would be best to only serialize the last value, but I've also added `Config#serialize` to simplify that use-case somewhat.

Haven-King committed:
Clarify that basic types include enums

Haven-King committed:
Add more type and null checks

Haven-King committed:
Clarify not nullness of config fields

CheaterCodes commented on src/main/java/org/quiltmc/loader/api/config/values/CompoundConfigValue.java:
```diff
@@ -0,0 +1,39 @@
+public interface CompoundConfigValue<T> {
+	Class<T> getType();
+
+	T getDefaultValue();
```
I guess the javadoc on `grow` explains that, but maybe another javadoc here would clear it up :)

CheaterCodes commented on src/main/java/org/quiltmc/loader/api/config/TrackedValue.java:
```diff
@@ -0,0 +1,215 @@
+	 *
+	 * @param newValue some value
+	 */
+	void setOverride(T newValue);
```
Even if the javadoc doesn't specify an intended use, it could at least explain the behaviour of the override (i.e., it's not being saved).

Haven-King committed:
Don't trigger callbacks if an override is present

Haven-King committed:
Clarify docs on CompoundConfigValue

Haven-King committed:
Clarify javadocs for overrides

CheaterCodes commented:
Another note on your opening comment:
> Why put this in Loader and not QSL?
You should specify on "other parts of loader": For now, it would probably be used for Chasm and loader plugins.
Another good reason is that non-minecraft mods get to use the same system.

Haven-King committed:
- Make Config.Builder extend Config.SectionBuilder
- Add and use ConfigParseException

Haven-King committed:
- Add some basic tests for complex values
- Fix some behavior for complex values

Haven-King committed:
- Widen Constrant.Range to be open to any types

Haven-King committed:
- Simplify callback invocation

CheaterCodes reviewed (changes requested):
Found an issue with missing directories.

CheaterCodes commented on src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java:
```diff
@@ -0,0 +1,239 @@
+	static TrackedValue<Boolean> TEST_BOOLEAN;
+	static TrackedValue<String> TEST_STRING;
+	static TrackedValue<ValueList<Integer>> TEST_LIST;
+
```
Actually, the proper way to do this is with a Temp dir provided by Junit

Haven-King replied:
Keeping the directory around is useful for validating results by hand and testing that deserialization is handled properly after manual changes to the config files.

CheaterCodes replied:
I only partially agree. Unit tests in a test suite should be consistent and not depend on the current state of the file system.
Manual testing should be done separately.

Haven-King committed:
- Remove flags
- Change metadata to hold a single value for each type
- Metadata is now created using a builder pattern

Haven-King committed:
- Clean up metadata methods and comments
- Abstract ConfigFieldAnnotationProcessor to apply to any MetadataContainerBuilder instead of just fields
- Allow for metadata (i.e. comments) on sections
- Modify NightConfigSerializer to preserve ordering

Haven-King committed:
- Fix issue with missing config directory

falkreon commented on src/main/java/org/quiltmc/loader/api/config/NightConfigSerializer.java:
```diff
@@ -0,0 +1,209 @@
+	}
+
+	@SuppressWarnings("unchecked")
+	private static Object coerce(Object object, Object to) {
```
Marshalling is a big can o' worms. This might want to be wholly set aside as its own class. It's kind of sad that java doesn't have builtin rules for it.

After marshalling to Classes and target objects, I've been moving very definitely towards marshalling to Types. You can extract a non-erased Type from a Field, and at that point you can follow the data hierarchy and stay in non-erased Types very easily. This also prevents you from having to construct default values for arbitrarily complex types.

Haven-King replied:
Could you perhaps give an example of marshalling using Types?

falkreon commented on src/main/java/org/quiltmc/loader/impl/config/Json5Serializer.java:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	@SuppressWarnings("unchecked")
+	private Object coerce(Object object, Object to) {
```
DRY problem - this is duplicate code

falkreon commented on src/main/java/org/quiltmc/loader/impl/config/tree/Trie.java:
```diff
@@ -0,0 +1,260 @@
+import org.quiltmc.loader.impl.config.builders.SectionBuilderImpl;
+import org.quiltmc.loader.impl.config.values.ValueKeyImpl;
+
+public final class Trie {
```
Needs javadoc to make it clear that this is a datastructure for efficient key or substring lookup, or whatever you're using it for.

falkreon commented on src/main/java/org/quiltmc/loader/impl/config/util/ConfigUtils.java:
```diff
@@ -0,0 +1,76 @@
+
+public final class ConfigUtils {
+	private static final Class<?>[] VALID_VALUE_CLASSES = new Class[] {
+			Integer.TYPE,
```
:heavy_check_mark: This will play nice with kotlin

falkreon reviewed (commented):
Still needs a little bit more javadoc, and I would deduplicate and harden marshalling code, but nothing jumps out to me as needing urgent changes.

Haven-King committed:
Make MetadataType a final class instead of an interface

Haven-King committed:
Extract marshalling function from the serializers to a common MarshallingUtils class

Haven-King committed:
Remove unused Trie method and add some clarifying javadoc

Haven-King committed:
Add space after matches constraint representation

falkreon commented:
:heavy_check_mark: fixes look great!

Haven-King committed:
Allow configs to contain arbitrary objects as long as those objects implement ComplexConfigValue. See Vec3i test.

Haven-King committed:
- Classes that represent config sections must now be explicitly designated as such
- Fix a small marshalling bug with nested maps

Haven-King committed:
Make tests use a temporary directory

Haven-King committed:
- Expose TrackedValue callback invoke and serialization methods
- Rename `fileType` to `format`

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Comments.java:
```diff
@@ -0,0 +1,4 @@
+package org.quiltmc.loader.api.config;
+
+public interface Comments extends Iterable<String> {
```
I have no idea what this class does just by looking at it.
Javadoc, please!

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	String getId();
+
+	/**
+	 * The path this config will be saved in, relative to Quilt Loader's config directory.
```
since we're going to make this separate from quilt loader
```suggestion
	 * The path this config will be saved in, relative to the root config directory.
```

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	Path getSavePath();
+
+	/**
+	 * Adds a listener to this {@link Config} that's called whenever any of its values is updated
```
```suggestion
	 * Adds a listener to this {@link Config} that's called whenever any of its values are updated
```

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	/**
+	 * Save this config and all its values to disk
+	 */
+	void serialize();
```
why not call it save(), then?

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	Iterable<TrackedValue<?>> values();
+
+	/**
+	 * @param key an iterable of key components to access
```
"key components" reads as "vital components", but i don't think that's what you intended here? Please re-word

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	TrackedValue<?> getValue(Iterable<String> key);
+
+	/**
+	 * @return all nodes of the value tree represented by this config file, including section nodes
```
You reference "value tree" a lot here, but I can't ever find a place that it's actually defined what that means. The javadoc for ValueTreeNode just says "An element in a config tree.", which is *even more confusing*, because now there's two phrases that are equally useless on describing what the "value/config tree" actually is, and what the difference between the two are

Haven-King replied:
I've added more docs regarding the behavior of this method that should help clarify things.

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+import org.quiltmc.loader.impl.config.ConfigImpl;
+
+@ApiStatus.NonExtendable
+public interface Config {
+	String getModId();
+
+	String getId();
```
javadoc!

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	 *         <li>An object whose class follows these rules</li>
+	 *     </ul>
+	 * </ul>
+	 *
```
It would probably be a good idea to note things that are never serialized, like (i presume) static fields, and fields with the `transient` keyword (which you should implement, by the way)

TheGlitch76 replied:
maybe mention the Comment annotation here?

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/Config.java:
```diff
@@ -0,0 +1,423 @@
+	 * @param modId the mod owning the resulting config file
+	 * @param id the config's id
+	 * @param path additional path elements to include as part of this configs file, e.g.
+	 *             if the path is empty, the config file might be ".minecraft/config/example_mod/id.json5"
+	 *             if the path is "client/gui", the config file might be ".minecraft/config/example_mod/client/gui/id.json5"
```
isn't the default toml now?

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/MetadataType.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.function.Supplier;
+
+/**
+ * A typed key to be used for setting and getting object metadata.
```
As far as I can tell, it's never explained what "object metadata" actually is.

Haven-King replied:
I think the idea of "metadata" in general is pretty clear, but I've added some more clarifying javadoc that hopefully helps to paint a better picture.

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/annotations/ConfigFieldAnnotationProcessor.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * <p>See {@link Comment}
+ */
+public interface ConfigFieldAnnotationProcessor<T extends Annotation> {
```
Why is this API?

TheGlitch76 reviewed (changes requested):
Here are some things I noticed reading through the API classes (except the values package). This is not a complete review, but I wanted to get these comments visible so that you have a chance to address them

Haven-King committed:
Update src/main/java/org/quiltmc/loader/api/config/Config.java

Co-authored-by: Glitch <glitch.g3431@gmail.com>

Haven-King committed:
Some javadoc additions and changes

Haven-King committed:
More javadocs

Haven-King committed:
Allow for multiple Comment annotations

Haven-King committed:
Update src/main/java/org/quiltmc/loader/api/config/Config.java

Co-authored-by: Glitch <glitch.g3431@gmail.com>

Haven-King committed:
Merge remote-tracking branch 'origin/master'

Haven-King committed:
More javadoc updates

Haven-King committed:
Clarify that transient fields aren't serialized (and make that so)

Haven-King committed:
More javadoc stuff

Haven-King committed:
- Rename some methods for conciseness
- Abstract Config#modId to Config#family for the imminent separation into a non Quilt Loader specific library

Haven-King committed:
Minor javadoc change

Haven-King committed:
Add javadocs pointing out the annotation system

Haven-King committed:
More minor method renames

Haven-King committed:
Minor javadoc change

Haven-King committed:
Add constraint annotations

Haven-King committed:
Oops duplicate javadoc

Haven-King committed:
Allow builtin constraint annotations to apply to CompoundConfigValues

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/NightConfigSerializer.java:
```diff
@@ -0,0 +1,155 @@
+import org.quiltmc.loader.api.config.values.ValueMap;
+import org.quiltmc.loader.api.config.values.ValueTreeNode;
+
+public final class NightConfigSerializer<C extends CommentedConfig> implements Serializer {
```
Thinking on this further--this should *not* be stable API, because it ties us to using NightConfig forever.

TheGlitch76 commented on src/main/java/org/quiltmc/loader/impl/config/Json5Serializer.java:
```diff
@@ -0,0 +1,248 @@
+
+	@Override
+	public void serialize(Config config, OutputStream to) throws IOException {
+		JsonWriter writer = JsonWriter.json5(new OutputStreamWriter(to));
```
Allowing strict json would be essentially free here (the only major concern would be to not write comments in strict mode), so it is probably worth implementing a JsonSerializer too?

Haven-King replied:
Even if it is free I really don't see a point to doing so.

CheaterCodes commented on src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java:
```diff
@@ -0,0 +1,245 @@
+	static TrackedValue<ValueList<Integer>> TEST_LIST;
+
+	public ConfigTester() {
+		QuiltLoaderImpl.INSTANCE.setGameProvider(new DummyGameProvider(TEMP));
```
Would it be possible to do this in a `@BeforeEach` method with `@TempDir Path conigDir` argument instead of the constructor? That way every test gets a fresh directory. Unless ofc this is difficult with the loader iomplementation.
And of course, this would then also suggest tests that write to an existing directory.

CheaterCodes commented on src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java:
```diff
@@ -0,0 +1,245 @@
+
+	@Test
+	public void testValidation() {
+		Assertions.assertThrows(RuntimeException.class, () -> {
```
I'm not a huge fan of expecting a RuntimeException. This will pass if there's *any* uncaught exception in the test. Would be nice to catch a more specific exception (with assertThrowsExactly as well).

CheaterCodes reviewed (commented):
Did you plan to pre-define a MetadataType for flags?

Haven-King committed:
Rename DiscreteRange and RealRange to IntegerRange and FloatRange

Haven-King committed:
Replace constructor and static temp initialization in tests

Haven-King committed:
Remove leftover flag stuff

Haven-King committed:
Move NightConfigSerializer to impl

Haven-King committed:
Fix system property for default config extension and add some comments

Haven-King commented on src/main/java/org/quiltmc/loader/impl/util/SystemProperties.java:
```diff
@@ -26,6 +26,8 @@ public final class SystemProperties {
 	public static final String MODS_DIRECTORY = "loader.modsDir";
 	public static final String CONFIG_DIRECTORY = "loader.configDir";
+	public static final String DEFAULT_CONFIG_EXTENSION = "loader.globalConfigExtension";
+	public static final String GLOBAL_CONFIG_EXTENSION = "loader.globalConfigExtension";
```
Yep, fixed, thanks

Haven-King committed:
Add and use a bunch of specific exception types

Haven-King committed:
Update tests to use more specific exception types

Haven-King committed:
Add some clarifying docs to the Matches constraint

Haven-King committed:
Replace reference to mod with family

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/Configs.java:
```diff
@@ -0,0 +1,45 @@
+	/**
+	 * @return all registered config files for the given mod
+	 */
+	public static Iterable<Config> getConfigs(String familyId) {
```
Family and mod ID are essentially interchangeable. I moved to using the term "family ID" for config families over mod ID since this config system will be its own library and may not be used specifically for mods.

CheaterCodes replied:
Family seems like an odd choice to me. Maybe `configGroup`?

Haven-King replied:
I don't think "group" is any more descriptive than "family". The term "family" is used in computing, such as font family for example.

CheaterCodes replied:
"Family" makes me think that they share some sort of structure or behavior. 

In the modding context, "family" would indicate to me grouping by function, i.e. "client family" or "rendering family". Group just feels more generic.

But yeah, this is mostly preference and I don't care too much.

Haven-King committed:
Move ConfigSerializers to API and move serializer entrypoint handling to Knot in preparation for the config system not depending on Loader.

CheaterCodes commented on src/main/java/org/quiltmc/loader/api/config/Constraint.java:
```diff
@@ -0,0 +1,139 @@
+		};
+	}
+
+	final class Range<T, BOUNDS> implements Constraint<T> {
```
If T is long/double an BOUNDS is int/float, what's stopping you from just up-casting the bounds?

TheGlitch76 commented on src/main/java/org/quiltmc/loader/impl/util/ImmutableIterable.java:
```diff
@@ -0,0 +1,47 @@
+
+import java.util.Iterator;
+
+public final class ImmutableIterable<T> implements Iterable<T> {
```
I don't understand how this makes the Iterable immutable? You never clone, just keep a reference to the inital object.

Haven-King replied:
`Iterator#remove` throws an `UnsupportedOperationException` by default. By wrapping whatever iterable is passed with a new subclass of `Iterator`, attempts to call `Iterator#remove` will result in throwing that exception.

TheGlitch76 commented on src/main/java/org/quiltmc/loader/impl/config/values/ValueMapImpl.java:
```diff
@@ -0,0 +1,190 @@
+import org.quiltmc.loader.api.config.values.ValueMap;
+import org.quiltmc.loader.impl.config.tree.TrackedValueImpl;
+
+public final class ValueMapImpl<T> implements ValueMap<T>, org.quiltmc.loader.api.config.values.CompoundConfigValue<T> {
```
uh, why is this fully qualified?

TheGlitch76 commented on src/main/java/org/quiltmc/loader/impl/config/values/ValueListImpl.java:
```diff
@@ -0,0 +1,313 @@
+import org.quiltmc.loader.api.config.values.ValueList;
+import org.quiltmc.loader.impl.config.tree.TrackedValueImpl;
+
+public final class ValueListImpl<T> implements ValueList<T>, org.quiltmc.loader.api.config.values.CompoundConfigValue<T> {
```
Why is this fully qualified?

TheGlitch76 reviewed (changes requested):
more comments

Haven-King commented on src/main/java/org/quiltmc/loader/api/config/Serializer.java:
```diff
@@ -0,0 +1,32 @@
+public interface Serializer {
+	String getFileExtension();
+
+	void serialize(Config config, OutputStream to) throws IOException;
+
+	void deserialize(Config config, InputStream from) throws IOException;
```
I don't think that would really provide much value given that these methods are each called in only two places and should only ever be called internally I really don't think having helper methods for paths adds much value.

Haven-King committed:
Remove qualifications

Haven-King committed:
Make ranges explicitly require their appropriate data types

Haven-King committed:
Restrict range types further

CheaterCodes reviewed (approved):
While I would like my comment to be addressed, it's really just preference anyway.
Last important thing that *needs* addressing is the comment by i5.

Apart from that, I'm ready for FCP, let's get this merged!

Haven-King committed:
Add some clarifying docs regarding metadata

Haven-King committed:
Add some clarifying javadoc on the config tree

Haven-King committed:
Remove ConfigWrapper, replacing it with a WrappedConfig class that config definition classes can extend to give direct access to Config methods.

TheGlitch76 commented on src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java:
```diff
@@ -0,0 +1,248 @@
+import org.quiltmc.loader.api.config.values.ValueList;
+import org.quiltmc.loader.impl.QuiltLoaderImpl;
+
+public class ConfigTester {
```
Unless I missed something, tests for `Matches` are needed here

TheGlitch76 commented on src/main/java/org/quiltmc/loader/api/config/values/ValueKey.java:
```diff
@@ -0,0 +1,36 @@
+import org.jetbrains.annotations.ApiStatus;
+
+@ApiStatus.NonExtendable
+public interface ValueKey extends Iterable<String> {
```
Unlike the other Value classes it's not immediately obvious to me what a "value key" is, so this is the last place I'm going to request some javadoc :pineapple~1:

TheGlitch76 reviewed (approved):
Some final notes, but nothing that you can't resolve before the FCP is over.

Haven-King committed:
Add some docs to ValueKey

Haven-King committed:
Add some tests for the Matches constraint

Haven-King commented:
Moved to final comment period. Ends June 6th.

osfanbuff63 reviewed (approved):
Looks pretty good, but you got i5's comment right? Unresolved so asking.

kvverti commented on src/main/java/org/quiltmc/loader/api/config/MetadataType.java:
```diff
@@ -0,0 +1,66 @@
+ * <p>See also {@link TrackedValue.Builder#metadata}, {@link Config.Builder#metadata}, and {@link Config.SectionBuilder#metadata}
+ */
+public final class MetadataType<T, B extends MetadataType.Builder<T>> {
+	private final Class<T> metadataClass;
```
May I recommend using `Type` instead of `Class<T>`? `Class<T>` doesn't support parameterized types, which makes `MetadataType<List<String>>` impossible without client-side unchecked operations.

Haven-King replied:
I chose to remove the class parameter altogether as it was not being used for anything.

Leo40Git commented on src/main/java/org/quiltmc/loader/api/config/Constraint.java:
```diff
@@ -0,0 +1,145 @@
+		public Optional<String> test(CompoundConfigValue<T> value) {
+			StringBuilder builder = new StringBuilder();
+
+			for (T t : value.values()) {
```
Maybe add the index of the erroring value here?

Haven-King replied:
I don't think this is necessary since each constraint specifies which values are failing their test and why.

Leo40Git commented on src/main/java/org/quiltmc/loader/impl/config/util/ConfigFieldAnnotationProcessors.java:
```diff
@@ -0,0 +1,60 @@
+import org.quiltmc.loader.impl.config.Comments;
+import org.quiltmc.loader.api.config.annotations.ConfigFieldAnnotationProcessor;
+
+public final class ConfigFieldAnnotationProcessors {
```
This is public API, so it should be moved to the API package?

Leo40Git reviewed (approved):
LGTM! Just 2 minor nitpicks.

AlexIIL commented on build.gradle:
```diff
@@ -80,6 +80,8 @@ dependencies {
 	include 'org.quiltmc:quilt-loader-sat4j:2.3.5.1'
 
+	api 'com.electronwill.night-config:core:3.6.5'
+	api 'com.electronwill.night-config:toml:3.6.5'
```
Like i5 said: we need this added to the `src/main/resources/quilt_installer.json` and `src/main/resources/fabric-installer.launchwrapper.json` files - I'm not completely sure where both are used, but it should be added to both to be safe.

Haven-King replied:
What are your and @TheGlitch76's thoughts on shadowing and relocating NightConfig rather than adding it to the dependency files? Since it is purely impl and not a library we control, this would allow us to avoid mirroring it on our Maven server.

AlexIIL replied:
I'd be okay with that too!

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/Comments.java:
```diff
@@ -0,0 +1,8 @@
+package org.quiltmc.loader.api.config;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/MarshallingUtils.java:
```diff
@@ -0,0 +1,112 @@
+package org.quiltmc.loader.api.config;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/MetadataContainerBuilder.java:
```diff
@@ -0,0 +1,14 @@
+package org.quiltmc.loader.api.config;
```
Missing Quilt License header

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/annotations/FloatRange.java:
```diff
@@ -0,0 +1,41 @@
+package org.quiltmc.loader.api.config.annotations;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/annotations/IntegerRange.java:
```diff
@@ -0,0 +1,41 @@
+package org.quiltmc.loader.api.config.annotations;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/annotations/Matches.java:
```diff
@@ -0,0 +1,39 @@
+package org.quiltmc.loader.api.config.annotations;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/exceptions/ConfigCreationException.java:
```diff
@@ -0,0 +1,18 @@
+package org.quiltmc.loader.api.config.exceptions;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/exceptions/ConfigFieldException.java:
```diff
@@ -0,0 +1,18 @@
+package org.quiltmc.loader.api.config.exceptions;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/exceptions/ConfigParseException.java:
```diff
@@ -0,0 +1,18 @@
+package org.quiltmc.loader.api.config.exceptions;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/exceptions/TrackedValueException.java:
```diff
@@ -0,0 +1,18 @@
+package org.quiltmc.loader.api.config.exceptions;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/values/ComplexConfigValue.java:
```diff
@@ -0,0 +1,18 @@
+package org.quiltmc.loader.api.config.values;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/api/config/values/ConfigSerializableObject.java:
```diff
@@ -0,0 +1,25 @@
+package org.quiltmc.loader.api.config.values;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/config/Comments.java:
```diff
@@ -0,0 +1,30 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL commented on src/main/java/org/quiltmc/loader/impl/config/CommentsImpl.java:
```diff
@@ -0,0 +1,23 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/ConfigTester.java:
```diff
@@ -0,0 +1,252 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Are you sure this should be a *fabric* licensed file? (It might make it easier to validate I suppose, but it looks like your original work?)

Haven-King replied:
It shouldn't, no. Unfortunately, the `licenseFormat` task only applies fabric license headers. That's also why most files don't have them, because once I realized this I just didn't bother futzing with the headers manually. When this PR is separated out into its own library, I will make sure all files have the appropriate Quilt headers.

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/DummyGameProvider.java:
```diff
@@ -0,0 +1,103 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/SerializationTester.java:
```diff
@@ -0,0 +1,74 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Are you sure this should be a *fabric* licensed file? (It might make it easier to validate I suppose, but it looks like your original work?)

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/TestValueListConfig.java:
```diff
@@ -0,0 +1,10 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/TestValueMapConfig.java:
```diff
@@ -0,0 +1,10 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL commented on src/test/java/org/quiltmc/loader/impl/config/Vec3i.java:
```diff
@@ -0,0 +1,34 @@
+package org.quiltmc.loader.impl.config;
```
Missing Quilt license header.

AlexIIL reviewed (changes requested):
Overall, this looks really good!

There are two main issues:
 * Some of your added files don't have the quilt license header - this should probably be fixed before we merge.
 * This fails to build (at all) in github actions. I'd be okay with temporarily "fixing" this by changing the github action task from `./gradlew build publish --stacktrace` to `./gradlew build publish --stacktrace -x checkLicense`, and opening an issue for us to fix this properly, since I don't want to delay merging this.

Haven-King committed:
Drop class parameter from MetadataType to allow for MetadataTypes with generic values.

Haven-King committed:
Move ConfigFieldAnnotationProcessors to API package

Haven-King committed:
Fix minor issue for compound constraint error message

Haven-King committed:
Extracts the majority of config API code out to a separate library

Haven-King committed:
Use system properties for config stuff

Leo40Git commented on src/main/java/org/quiltmc/loader/impl/config/QuiltConfigImpl.java:
```diff
@@ -30,17 +34,45 @@ private QuiltConfigImpl() {
+		}
+
+		if (defaultConfigExtension != null && !serializerMap.containsKey(defaultConfigExtension)) {
+			throw new RuntimeException("Cannot use file extension " + globalConfigExtension + " globally: no matching serializer found");
```
```suggestion
			throw new RuntimeException("Cannot use file extension " + defaultConfigExtension + " by default: no matching serializer found");
```

Haven-King committed:
- Shadow NightConfig
- Add QuiltConfig to installer files

Haven-King committed:
Fix error message

Co-authored-by: ADudeCalledLeo <7997354+Leo40Git@users.noreply.github.com>

triphora commented on src/main/java/net/fabricmc/loader/launch/knot/Knot.java:
```diff
@@ -153,6 +154,8 @@ protected ClassLoader init(String[] args) {
 		unlocked = true;
 
 		try {
+			QuiltConfigImpl.init();
```
Why is this in the try/catch?

Haven-King committed:
Move QuiltConfig initialization outside of prelaunch try/catch

Haven-King committed:
Merge remote-tracking branch 'origin/master'

Haven-King committed:
Update QuiltConfig version

Haven-King committed:
Update QuiltConfig version

Haven-King committed:
Merge branch 'develop' of https://github.com/QuiltMC/quilt-loader
